hushline/crypto/secrets_manager.py

```diff
@@ -0,0 +1,113 @@
+"""
+A class definition to simplify the careful handling of cryptographic secrets.
+"""
+
+import warnings
+from base64 import b64decode, urlsafe_b64encode
+from hashlib import shake_256
+from pathlib import Path
+from secrets import token_bytes
+
+from aiootp.generics.canon import canonical_pack, fullblock_ljust
+from cryptography.fernet import Fernet, InvalidToken
+from passlib.hash import argon2
+
+__all__ = ["InvalidToken", "SecretsManager"]
+
+
+def truncated_b64decode(value: bytearray) -> bytearray:
+    if bool(remainder := len(value) % 4):
+        value.extend((4 - remainder) * b"=")
+    return bytearray(b64decode(value))
+
+
+class SecretsManager:
+    __slots__ = ("_secret_salt_filename", "_secret_salt_length", "_memory_cost", "_kdf")
+
+    _APP_SECRETS_DIRECTORY: Path = Path("hushline/crypto/_app_secrets")
+    _KDF_BLOCKSIZE: int = shake_256().block_size
+    _MEMORY_COST_IN_KiB: int = 128 * 1024
+    _SECRET_SALT_FILENAME: str = "_device_salt.txt"
+    _SECRET_SALT_LENGTH: int = 32
+
+    def __init__(
+        self,
+        admin_secret: bytearray,
+        *,
+        memory_cost: int | None = None,
+        secret_salt_filename: str | None = None,
+        secret_salt_length: int | None = None,
+    ) -> None:
+        self._APP_SECRETS_DIRECTORY.mkdir(parents=True, exist_ok=True)
+        self._memory_cost = self._MEMORY_COST_IN_KiB if memory_cost is None else memory_cost
+        self._secret_salt_filename = (
+            self._SECRET_SALT_FILENAME if secret_salt_filename is None else secret_salt_filename
+        )
+        self._secret_salt_length = (
+            self._SECRET_SALT_LENGTH if secret_salt_length is None else secret_salt_length
+        )
+        self._prepare_key_derivation_object(admin_secret)
+
+    def _summon_device_salt(self) -> bytearray:
+        salt_path = self._APP_SECRETS_DIRECTORY / self._secret_salt_filename
+
+        if not salt_path.is_file():
+            salt_path.write_bytes(salt := token_bytes(self._secret_salt_length))
```

Comment:
Hm, the local filesystem isn't available in some of the intended environments... It would be preferred to keep it out of the database... But, if it's the only commonly available location, it may have to reside there.

@jeremywmoore @micahflee Would there be any technical issues preventing the salt from being stored in the database?

Reply:
I don't see a technical reason why we couldn't.

Reply:
Okay, I'll pivot towards using the database for initialization values.

Reply:
Addressed by (https://github.com/scidsg/hushline/pull/411/commits/294338bb20d60ed5a2900f842e1fb432460d5d3a)